### PR TITLE
Add Support for Restoring Specific Backups

### PR DESCRIPTION
**Pull request by mattlord**
## Description
By default Vitess will only make practical use of the latest backup of a given shard. While this makes perfect sense for the common use cases there are times where you may want to restore a specific backup. For example:
  1. In order to extract a portion of the data that can then be merged with the current state. For example if you later realize that you accidentally deleted some records in a table that you shouldn't have last week, and you need to perform a restore so that you can copy those specific records back to the live data set.
  2. To perform validation, forensics, analysis on the system state at that time.
  3. A specific PITR for whatever reason ...

This work supports that in two different ways:
1. You can start a tablet with: `-restore_from_backup -restore_from_backup_ts 2021-04-29.133050`
2. You can use vtctl with: `vtctlclient -server=<vtctld-server>:<vtctld-port> RestoreFromBackup -backup_timestamp=2021-04-29.133050 <tablet-alias>`

### This compared with PITR
While Vitess supports a method to accomplish [PITR](https://vitess.io/docs/reference/features/recovery/), it's a fairly involved process with the details left up to the user (e.g. binlog servers). This work offers another method that can potentially be used in various ways/circumstances as an alternative. For example, let's say we realize that 1-2 days ago we made a mistake and want to look at the previous state of the data on one tablet in the shard... we could do that with:
```
$ vtctlclient -server=<vtctld-server>:<vtctld-port> RestoreFromBackup \
    -backup_timestamp=$(date -d"2 days ago" +"%Y-%m-%d").000000 <tablet-alias> && \
      vtctlclient -server=<vtctld-server>:<vtctld-port> StopReplication <tablet-alias>
```

This will stop MySQL replication on the tablet ASAP after the restore completes and prevent it from being automatically re-started (via tablet repair). This allows you to do whatever you like on that tablet before deciding to later:
1. Throw it away (perhaps we added this new tablet just for this purpose)
2. Let it catch up with replication normally (just starting replication again via `vtctlclient StartReplication`)
3. Perform a Restore on it using the latest shard backup
4. ...

Co-authored-by: Guido Iaquinti <giaquinti@slack-corp.com>
Signed-off-by: Matt Lord <mattalord@gmail.com>

## Related Issue(s)
This is a continuation of: https://github.com/vitessio/vitess/pull/7998
This solves: https://github.com/vitessio/vitess/issues/4905

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [x] Documentation is needed: https://github.com/vitessio/website/pull/834

### CODE REVIEW

**Comment by guidoiaquinti** [comment]
Once tests are added LGTM. Thank you for moving this forward!

**Comment by mattlord** [comment]
I've tested and verified the tablet flag as well as the vtctlclient flag:
# Backups
```
$ vtctlclient ListBackups commerce/0
2021-09-17.023140.zone1-0000000101
2021-09-17.023359.zone1-0000000101
```

# Updated help output
```
$ vtctlclient RestoreFromBackup --help
Usage: RestoreFromBackup [-backup_timestamp=yyyy-mm-dd.HHmmss] <tablet alias>

Stops mysqld and restores the data from the latest backup or if a timestamp is specified then the most recent backup at or before that time.

  -backup_timestamp string
    	Use the backup taken at or before this timestamp rather than using the latest backup.
```

# Invalid timestamp
```
$ vtctlclient RestoreFromBackup -backup_timestamp=222 zone1-101
RestoreFromBackup Error: rpc error: code = Unknown desc = TabletManager.RestoreFromBackup on zone1-0000000101 error: unable to parse the timestamp specified for -restore_from_backup_ts of 222: parsing time "222" as "2006-01-02.150405": cannot parse "222" as "2006": unable to parse the timestamp specified for -restore_from_backup_ts of 222: parsing time "222" as "2006-01-02.150405": cannot parse "222" as "2006"
E0917 02:24:02.288281    2671 main.go:76] remote error: rpc error: code = Unknown desc = TabletManager.RestoreFromBackup on zone1-0000000101 error: unable to parse the timestamp specified for -restore_from_backup_ts of 222: parsing time "222" as "2006-01-02.150405": cannot parse "222" as "2006": unable to parse the timestamp specified for -restore_from_backup_ts of 222: parsing time "222" as "2006-01-02.150405": cannot parse "222" as "2006"
```

# Uses latest / 2021-09-17.023359
```
$ vtctlclient RestoreFromBackup zone1-101
I0917 02:37:45.903227    3273 main.go:67] I0917 02:37:45.900682 backup.go:264] I0917 02:37:45.900366 backup.go:270] Restore: looking for a suitable backup to restore
I0917 02:37:45.903530    3273 main.go:67] I0917 02:37:45.901729 backup.go:264] I0917 02:37:45.900947 backupengine.go:221] Restore: found backup commerce/0 2021-09-17.023359.zone1-0000000101 to restore using a timestamp of 0001-01-01.000000
...
I0917 02:37:53.528038    3273 main.go:67] I0917 02:37:53.527748 backup.go:264] I0917 02:37:53.527519 backup.go:356] Restore: restarting mysqld after mysql_upgrade
```

# Uses the older one based on the timestamp
```
$ vtctlclient RestoreFromBackup -backup_timestamp=2021-09-17.023140 zone1-101
I0917 02:39:13.405321    4445 main.go:67] I0917 02:39:13.404759 backup.go:264] I0917 02:39:13.404568 backup.go:270] Restore: looking for a suitable backup to restore
I0917 02:39:13.407096    4445 main.go:67] I0917 02:39:13.406871 backup.go:264] I0917 02:39:13.406730 backupengine.go:221] Restore: found backup commerce/0 2021-09-17.023140.zone1-0000000101 to restore using a timestamp of 2021-09-17.023140
...
I0917 02:39:21.234004    4445 main.go:67] I0917 02:39:21.233826 backup.go:264] I0917 02:39:21.233596 backup.go:356] Restore: restarting mysqld after mysql_upgrade
```

Now I have a good idea of what to test and how in the test suite. 🙂

File: go/vt/proto/tabletmanagerdata/tabletmanagerdata.pb.go
Comment (deepthi) [inline go/vt/proto/tabletmanagerdata/tabletmanagerdata.pb.go:4458]: PR is missing the protobuf changes from which this is generated.

**Review by deepthi — COMMENTED** [review]
Looking good, except protobuf definition is missing.

File: go/vt/proto/tabletmanagerdata/tabletmanagerdata.pb.go
Comment (mattlord) [reply go/vt/proto/tabletmanagerdata/tabletmanagerdata.pb.go:4458]: Good catch! Thank you. Added it.

**Review by deepthi — APPROVED** [review]
LGTM.
Let's get another set of eyes before merging.
Can you please create a website PR to document the new usage (vttablet flag & vtctl command)?

**Comment by mattlord** [comment]
> Can you please create a website PR to document the new usage (vttablet flag & vtctl command)?

https://github.com/vitessio/website/pull/834 (noted in the description now and linked in comments)

File: go/vt/vttablet/tabletmanager/restore.go
Comment (ajm188) [inline go/vt/vttablet/tabletmanager/restore.go:165]: I skimmed the rest of the PR and it doesn't seem like anyone brought this up. Another way to do this would be to push the parsing further up to the cli, and enforce stricter types closer to this function. In this way we could:
* `commandRestoreBackup` in `go/vt/vtctl/vtctl.go` parses its `backupTimestampStr` into a `time.Time`, using `mysqlctl.BackupTimestampFormat`
* we pass `time.Time` to all the tmc functions
* when making the gRPC request, we change that protobuf definition to be a `vttime.Time`, and call `protoutil.TimeToProto` to convert the `time.Time` to a `vttime.Time`, and convert back on the tabletmanager side.

What do you think?

**Review by ajm188 — APPROVED** [review]
I wanted to bring up one comment about the types we're using, but if you don't feel it's worth it to change then I'm happy to merge as-is.

File: go/vt/vttablet/tabletmanager/restore.go
Comment (mattlord) [reply go/vt/vttablet/tabletmanager/restore.go:165]: I briefly considered this but wasn't sure of the benefits (avoiding the RPC call with an invalid vtctl CLI parameter is one). I don't mind making this change if you prefer it? Sounds like you do or you wouldn't have mentioned it 😄 ... so I'll work on that now. Thanks!

Comment (mattlord) [reply go/vt/vttablet/tabletmanager/restore.go:165]: OK, after trying it again I'm remembering WHY I didn't go this route last time. 🙂

What makes it feel awkward/clunky to use `time.Time` coming from the `vtctl` side is that there's also a corresponding `vttablet` flag that's used directly in the tmc: https://github.com/vitessio/vitess/blob/a7a951beeee056af478958f233c81dbe417c33fb/go/vt/vttablet/tabletmanager/restore.go#L49

As things are now, it's uniform handling (always a string) on the tmc side that we convert when using -- whether the value came from a vtctl call and RPC or from a vttablet flag. Maybe I'm missing something though?

I understand the pull to use a more precise type... but it also feels like it makes the code more complex with little practical benefit. If you feel strongly about it though I can continue.

Comment (mattlord) [reply go/vt/vttablet/tabletmanager/restore.go]: OK, I just switched to this method. Thank you for the nudge! I feel better about it too and It wasn't really that awkward in the end. :-)

**Review by ajm188 — APPROVED** [review]
lgtm!

**Pull request merged**
